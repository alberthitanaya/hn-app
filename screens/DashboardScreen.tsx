import React, { useEffect, useState } from "react";
import { FlatList, SafeAreaView, StyleSheet } from "react-native";
import { CoinListItem } from "../components/CoinListItem/CoinListItem";
import balancesJson from "../constants/balances.json";
import OverallPosition from "../components/OverallPosition/OverallPositionContainer";
import { RootTabScreenProps } from "../types";
import { useNavigation } from "@react-navigation/core";
import AsyncStorage from "@react-native-async-storage/async-storage";
import "intl";
import { coingeckoUrl } from "../constants/Api";

interface GraphValues {
  timestamp: number;
  value: number;
}
const currenciesIdsCSV =
  "uniswap,dogecoin,terra-luna,ethereum,safemoon,litecoin,dash";

export default function DashboardScreen({
  navigation,
}: RootTabScreenProps<"Dashboard">) {
  const [data, setData] = useState<GraphValues[]>();
  const [balances, setBalances] = useState<any>();
  const [currentCurrency, setCurrentCurrency] = useState("aud");
  const [isRefreshing, setIsRefreshing] = useState<any>(false);
  const { navigate } = useNavigation();

  const fetchData = async () => {
    try {
      setIsRefreshing(true);
      const currency = (await AsyncStorage.getItem("currency")) || "aud";
      const url = `${coingeckoUrl}simple/price?vs_currencies=USD,AUD&include_24hr_vol=true&include_market_cap&include_24hr_change=true&&ids=${currenciesIdsCSV}`;
      setCurrentCurrency(currency);
      const formatter = new Intl.NumberFormat("en-US", {
        currency,
        style: "currency",
      });
      const response = await fetch(url);
      const json = await response.json();
      const currentBalances = balancesJson[
        balancesJson.length - 1
      ].balancesByCurrency.map((bal) => {
        return {
          ...bal,
          change: json[bal.currency.id][`${currency}_24h_change`],
          vol: formatter.format(json[bal.currency.id][`${currency}_24h_vol`]),
          price: formatter.format(json[bal.currency.id][currency || "aud"]),
          totalValue: formatter.format(
            bal.balance * json[bal.currency.id][currency || "aud"]
          ),
        };
      });
      setBalances(currentBalances);

      // get todays price
      const data = balancesJson.map((period) => {
        const result = period.balancesByCurrency.map((balance) => {
          return balance.balance * json[balance.currency.id][currency];
        });
        const change24hr = period.balancesByCurrency.map((balance) => {
          return (
            balance.balance *
            json[balance.currency.id][`${currency}_24h_change`]
          );
        });
        const reducer = (prev: number, curr: number) => prev + curr;
        // sum coins value
        const sum = result.reduce(reducer);
        // sum coins change in the last 24hr
        const change = change24hr.reduce(reducer);
        return {
          change,
          timestamp: period.timestamp,
          value: sum,
        };
      });
      setData(data);
      setIsRefreshing(false);
    } catch (error) {
      setIsRefreshing(false);
      alert("An network error has occurred, please refresh to try again");
      console.log("error", error);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    const unsubscribe = navigation.addListener("focus", () => {
      // The screen is focused
      // Call any action
      fetchData();
    });

    // Return the function to unsubscribe from the event so it gets removed on unmount
    return unsubscribe;
  }, [navigation]);

  return (
    <SafeAreaView testID="DashboardScreen" style={styles.container}>
      <FlatList
        accessibilityLabel="DashboardScreen"
        refreshing={isRefreshing}
        onRefresh={() => {
          fetchData();
        }}
        data={balances}
        keyExtractor={(item) => item?.item?.currency.id}
        showsVerticalScrollIndicator={false}
        renderItem={(item) => {
          return (
            <CoinListItem
              item={item}
              onPress={() =>
                navigate("CoinDetail", { id: item.item.currency.id })
              }
            />
          );
        }}
        ListHeaderComponent={
          <OverallPosition data={data} currency={currentCurrency} />
        }
      />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 70,
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: "80%",
  },
});
