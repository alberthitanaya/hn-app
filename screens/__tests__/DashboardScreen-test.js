import React from "react";
import renderer from "react-test-renderer";

import DashboardScreen from "../NotFoundScreen";

// const createTestProps = (props: Object) => ({
//     navigation: {
//       navigate: jest.fn()
//     },
//     ...props
//   });

const mockedDispatch = jest.fn();

// Mocks like this need to be configured at the top level 
// of the test file, they can't be setup inside your tests.
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: mockedDispatch,
    }),
  };
});
describe("<App />", () => {
  beforeEach(() => {
    // Alternatively, set "clearMocks" in your Jest config to "true"
    mockedDispatch.mockClear();
  });

  it("has 1 child", () => {
    const tree = renderer.create(<DashboardScreen ></DashboardScreen>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
