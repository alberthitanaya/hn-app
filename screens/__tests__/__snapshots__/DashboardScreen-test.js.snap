// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> has 1 child 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        },
      ]
    }
  >
    This screen doesn't exist.
  </Text>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginTop": 15,
        "opacity": 1,
        "paddingVertical": 15,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "color": "#2e78b7",
            "fontSize": 14,
          },
        ]
      }
    >
      Go to home screen!
    </Text>
  </View>
</View>
`;
